src/Microsoft.OData.Core/Json/ODataJsonWriterUtils.cs

```diff
@@ -213,31 +213,39 @@ internal static void EndJsonPaddingIfRequired(IJsonWriter jsonWriter, ODataMessa
         /// <param name="innerErrorPropertyName">The property name for the inner error property.</param>
         /// <param name="recursionDepth">The number of times this method has been called recursively.</param>
         /// <param name="maxInnerErrorDepth">The maximum number of nested inner errors to allow.</param>
+        /// <param name="skipNullProperties">if set, stacktrace, type, and message will not be written if they are empty strings on ODataInnerError.</param>
+
         private static void WriteInnerError(IJsonWriter jsonWriter, ODataInnerError innerError, string innerErrorPropertyName, int recursionDepth, int maxInnerErrorDepth)
         {
             Debug.Assert(jsonWriter != null, "jsonWriter != null");
             Debug.Assert(innerErrorPropertyName != null, "innerErrorPropertyName != null");
 
             ValidationUtils.IncreaseAndValidateRecursionDepth(ref recursionDepth, maxInnerErrorDepth);
 
-            // "innererror": {
+            // "innererror":
             jsonWriter.WriteName(innerErrorPropertyName);
             jsonWriter.StartObjectScope();
 
-            //// NOTE: we add empty elements if no information is provided for the message, error type and stack trace
-            ////       to stay compatible with Astoria.
-
-            // "message": "<message>"
-            jsonWriter.WriteName(JsonConstants.ODataErrorInnerErrorMessageName);
-            jsonWriter.WriteValue(innerError.Message ?? string.Empty);
-
-            // "type": "<typename">
-            jsonWriter.WriteName(JsonConstants.ODataErrorInnerErrorTypeNameName);
-            jsonWriter.WriteValue(innerError.TypeName ?? string.Empty);
-
-            // "stacktrace": "<stacktrace>"
-            jsonWriter.WriteName(JsonConstants.ODataErrorInnerErrorStackTraceName);
-            jsonWriter.WriteValue(innerError.StackTrace ?? string.Empty);
+            if (innerError.Properties != null)
+            {
+                foreach (KeyValuePair<string, ODataValue> pair in innerError.Properties)
+                {
+                    jsonWriter.WriteName(pair.Key);
+
+                    if ((pair.Key == JsonConstants.ODataErrorInnerErrorMessageName ||
+                        pair.Key == JsonConstants.ODataErrorInnerErrorStackTraceName ||
+                        pair.Key == JsonConstants.ODataErrorInnerErrorTypeNameName) &&
+                        pair.Value is ODataNullValue)
+                    {
+                        // Write empty string for null values in stacktrace, type and message properties of inner error. 
+                        jsonWriter.WriteJsonValue(string.Empty);

```

Comment:
>jsonWriter.WriteJsonValue(string.Empty); [](start = 24, length = 40)

Could you instead call WriteODataValue(new ODataStringValue(string.Empty)), and leave WriteJsonValue private?